### DISCUSSION TITLE
Does cloud-hypervisor should support Centos8.4 guest image?

### DISCUSSION THREAD

**Discussion by ufo19890607** [cloud-hypervisor/cloud-hypervisor#4673] [General]
I just want to use Centos8.4 with kernel 5.10. And I am sure that it can work with the same kernel version with Ubuntu image. 

To Reproduce

```
#!/bin/sh
vmlinux=$1
image=$2
./cloud-hypervisor/release/cloud-hypervisor \
        --kernel ./vmlinux/$vmlinux \
        --console off \
        --serial tty \
        --disk path=./images/$image \
        --cmdline "console=ttyS0 root=/dev/vda1 rw" \
        --cpus boot=4 \
        --memory size=1024M \
        --net "tap=,mac=,ip=,mask=" \
```
```
[root@VM-0-4-centos rust_vmm]# sh run-vm.sh vmlinux-5.10.25-85 CentOS-8-ec2-8.4.2105-20210603.0.x86_64.raw
[    0.452757] /dev/root: Can't open blockdev
[    0.453228] VFS: Cannot open root device "vda1" or unknown-block(254,1): error -6
[    0.454065] Please append a correct "root=" boot option; here are the available partitions:
[    0.454978] 0100           16384 ram0 
[    0.454980]  (driver?)
[    0.455664] 0101           16384 ram1 
[    0.455665]  (driver?)
[    0.456348] 0102           16384 ram2 
[    0.456348]  (driver?)
[    0.457031] 0103           16384 ram3 
[    0.457032]  (driver?)
[    0.457725] 0104           16384 ram4 
[    0.457726]  (driver?)
[    0.458408] 0105           16384 ram5 
[    0.458409]  (driver?)
[    0.459093] 0106           16384 ram6 
[    0.459093]  (driver?)
[    0.459777] 0107           16384 ram7 
[    0.459778]  (driver?)
[    0.460459] 0108           16384 ram8 
[    0.460460]  (driver?)
[    0.461146] 0109           16384 ram9 
[    0.461146]  (driver?)
[    0.461840] 010a           16384 ram10 
[    0.461841]  (driver?)
[    0.462537] 010b           16384 ram11 
[    0.462542]  (driver?)
[    0.463234] 010c           16384 ram12 
[    0.463235]  (driver?)
[    0.463927] 010d           16384 ram13 
[    0.463928]  (driver?)
[    0.464621] 010e           16384 ram14 
[    0.464622]  (driver?)
[    0.465311] 010f           16384 ram15 
[    0.465312]  (driver?)
[    0.466013] fe00        10485760 vda 
[    0.466015]  driver: virtio_blk
[    0.466772] fc00       262144000 nullb0 
[    0.466773]  (driver?)
[    0.467473] 103:00000   10485760 pmem0 
[    0.467474]  driver: nd_pmem
[    0.468223] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(254,1)
[    0.469179] Kernel Offset: disabled
[    0.469569] ---[ end Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(254,1) ]---
```
It can work with Ubuntu

```
[root@VM-0-4-centos rust_vmm]# sh run-vm.sh vmlinux-5.10.25-85 focal-server-cloudimg-amd64.raw
```

**Comment by iggy** [comment]
Are you sure that CentOS image is raw? It looks like the kernel isn't reading the partition table (note vda, but no vda1/2/etc). Does that image work with any other vmm (qemu/firecracker/etc)?

**Reply by ufo19890607** [reply]
hi, iggy,
I'm sure that Centos8.4 image is raw.
```
[root@VM-0-4-centos images]# qemu-img info CentOS-8-ec2-8.4.2105-20210603.0.x86_64.raw
image: CentOS-8-ec2-8.4.2105-20210603.0.x86_64.raw
file format: raw
virtual size: 10 GiB (10737418240 bytes)
disk size: 2.27 GiB
[root@VM-0-4-centos images]# 
```
I'm also sure that qemu can work well for that image.

**Reply by ufo19890607** [reply]
I think you can have a try for centos8.4 image. 

https://cloud.centos.org/centos/8/x86_64/images/

**Comment by rbradford** [comment]
You need a kernel with `CONFIG_MSDOS_PARTITION=y` and `CONFIG_XFS_FS=y` in the config. I have no idea where you got your kernel from but it definitely doesn't have the former in since it doesn't see the partition. The Cloud Hypervisor reference 5.15 kernel doesn't have either of these turned on (I compiled a kernel from the CH reference kernel with both those turned on and was able to boot.)

**Reply by rbradford** [reply]
And ordinarily I would recommend that you boot an image using a firmware so you don't have to worry about kernel config however unfortunately RedHat still don't ship EFI based "cloud" images for CentOS so neither Rust Hypervisor Firmware nor our port of TianoCore/OVMF/edk2 works.

**Reply by cryptoknight** [reply]
The UEFI variant of the AlmaLinux cloud image boots fine under rust firmware if you ever need a RHEL clone to test, FYI: https://wiki.almalinux.org/cloud/Generic-cloud.html

**Reply by lobshunter** [reply]
AlmaLinux cloud image works perfectly, however it doesn't provide older versions now(v8.4 for example). Any advice to run a RedHat:8.4 compatible clone?

Building from a packer file like [AlmaLinux does](https://github.com/AlmaLinux/cloud-images) might work. I am trying to find an easier way before working on it.
